### DISCUSSION TITLE
Applying a date formatter to a date field

### DISCUSSION THREAD

**Discussion by brandonpittman** [adobe/react-spectrum#4322] [Q&A] [answered]
I tried passing a date formatter like 

```ts
let dateFormatter = useDateFormatter({ month: "2-digit", day: "2-digit" })
```

to `useDateField` in the `state` object, but the resulting date input doesn't force two digit months or days. Is there a way to make `useDateField` create an input like that?

**Comment by snowystinger** [comment]
Seems like something we could expose on the options here: https://github.com/adobe/react-spectrum/blob/main/packages/%40react-stately/datepicker/src/useDateFieldState.ts#L190

At least, doing that works for me locally. It probably didn't work for you since the internals of useDateFieldState would need access to the same formatter you've created.

@devongovett any concerns?

**Reply by devongovett** [reply]
See https://github.com/adobe/react-spectrum/issues/4247#issuecomment-1481370983

**Reply by snowystinger** [reply]
my understanding of that comment was that it had more to do with an override as opposed to an option to the formatter itself
```
new Intl.DateTimeFormat('en-US', {month: "2-digit", day: "2-digit", year: 'numeric'}).format(new Date())
> '04/03/2023'
new Intl.DateTimeFormat('fr-FR', {month: "2-digit", day: "2-digit", year: 'numeric'}).format(new Date())
> '03/04/2023'
```

**Reply by devongovett** [reply]
Yes, the concern is that this override is not expected in many locales. IMO it's better to just let the formatter handle it based on the locale.

**Reply by brandonpittman** [reply]
The `ja-JP` defaults to a date like `2023/3/3`, but the design I've been tasked with requires a display of `2023/03/03`. You can use `Intl.DateTimeFormat` to make that date, but you can't represent it in `DateField`.

**Reply by snowystinger** [reply]
Can you expand on your business reason for needing this? I'm talking to our internationalization team and while this may not be in and of itself confusing, it might create inconsistencies leading to a different kind of confusion.

**Reply by brandonpittman** [reply]
I was able to sell my designer on not needing this so it’s a non-issue now.

**Reply by simenbrekken-visma** [reply, marked as answer]
I just came across this issue myself where we support multiple locales but dates are always displayed in ISO-8601 format. A somewhat hacky way to accomplish this is by mapping `DateSegments[]` to a different list of segments:

```ts
const state = useDateFieldState(...);
const segments = getDateFieldSegments(state.segments);


function getDateFieldSegments(segments: DateSegment[]): DateSegment[] {
    const daySegment = segments.find((segment) => segment.type === 'day');
    const monthSegment = segments.find((segment) => segment.type === 'month');
    const yearSegment = segments.find((segment) => segment.type === 'year');
    const separatorSegment = segments.find(
        (segment) => segment.type === 'literal'
    );

    if (!daySegment || !monthSegment || !yearSegment || !separatorSegment) {
        throw new Error('Invalid date field segments');
    }

    const formattedDaySegment = {
        ...daySegment,
        text: daySegment.text.padStart(2, '0'),
    };
    const formattedMonthSegment = {
        ...monthSegment,
        text: monthSegment.text.padStart(2, '0'),
    };
    const formattedSeparatorSegment = { ...separatorSegment, text: '-' };

    return [
        yearSegment,
        formattedSeparatorSegment,
        formattedMonthSegment,
        formattedSeparatorSegment,
        formattedDaySegment,
    ];
}

```

This will convert any format to "YYYY-MM-DD".

**Reply by RIP21** [reply]
Will it be interactive? Like with arrow keys etc?

**Reply by simenbrekken-visma** [reply]
It behaves exactly as it did before, just reordered.
